### PR TITLE
[store] Change database and table operations to kv

### LINKED ISSUE #1080: [store] impl sled db backed raft-state-machine.
- [x] #1311
- [x] #1312
- [x] #1313
- [ ] #1139 
- [x] #1385
- [x] #1386

### LINKED ISSUE #1286: [store] Make database and table operations delegate to KV
**Summary**

Currently, the database and table operations are independent of kv, we need to delegate these operations to kv to achieve a unified storage model to make the statemachine structure uniform.

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

Change database and table operations to kv.

## Changelog

- Improvement

## Related Issues

Fixes #1286 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -40,6 +40,7 @@ stateless-cluster-test-tls:
 	bash ./scripts/ci/ci-run-stateless-tests-cluster-tls.sh
 
 test: unit-test stateless-test
+	$(info ************ All tests passed! You are lucy!  ************)
```
Comment (jyizheng) [inline Makefile:43]: lucky?

**Comment by drmingdrmer** [comment]
For non-published PR(e.g., a draft), `rebase` would be better for sync-ing upstream changes. It results in a clean history.

A published PR prefers `merge` because a merge keeps complete history thus a pr comment will never lose its position.

**Comment by ariesdevil** [comment]
@drmingdrmer @dantengsky 
This implementation leaves data_parts behind, as that part of the design is not yet complete. I want to keep up with master first, sorry for the long delay, and thanks for your patience.

Also, current implementations of database and table operations involve fetch-modify-update operations on kv store(e.g. one table op includes two or more access to kv) that I haven't figured out if I need to implement a redis-like hash structure to improve the current functionality.

**Comment by dantengsky** [comment]
> @drmingdrmer @dantengsky
> This implementation leaves data_parts behind, as that part of the design is not yet complete. I want to keep up with master first, sorry for the long delay, and thanks for your patience.
> 
> Also, current implementations of database and table operations involve fetch-modify-update operations on kv store(e.g. one table op includes two or more access to kv) that I haven't figured out if I need to implement a redis-like hash structure to improve the current functionality.

It is not a trivial refactoring task, really appreciate it.

**Comment by drmingdrmer** [comment]
> Also, current implementations of database and table operations involve fetch-modify-update operations on kv store(e.g. one table op includes two or more access to kv) that I haven't figured out if I need to implement a redis-like hash structure to improve the current functionality.

Do not worry about it yet.
Soon later all of the storage will be migrated to sled. Then the concurrency control can be done with in a `sled::Transaction`.

**Comment by ariesdevil** [comment]
> > Also, current implementations of database and table operations involve fetch-modify-update operations on kv store(e.g. one table op includes two or more access to kv) that I haven't figured out if I need to implement a redis-like hash structure to improve the current functionality.
> 
> Do not worry about it yet.
> Soon later all of the storage will be migrated to sled. Then the concurrency control can be done with in a `sled::Transaction`.

OK, when the ops within the transaction, current `Database` struct should change. In this impl the database ops are serde simply, within transaction it should be db_name->db_id, then use db_id as the prefix to fetch all tables belong to this database. The operation of the table and data parts are similar.

File: fusestore/store/src/lib.rs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2020-2021 The Datafuse Authors.
 //
 // SPDX-License-Identifier: Apache-2.0.
-
+#![feature(type_ascription)]
 #[allow(clippy::all)]
 pub mod protobuf {
```
Comment (drmingdrmer) [inline fusestore/store/src/lib.rs:6]: 🤔 what does `type_ascription` do here?

**Comment by ariesdevil** [comment]
@PsiACE I want to know how to avoid automatically closing the PR when I change the origin branch name? I remember you did a branch name change before.

**Review by drmingdrmer — APPROVED** [review]
LGTM

**Comment by PsiACE** [comment]
> @PsiACE I want to know how to avoid automatically closing the PR when I change the origin branch name? I remember you did a branch name change before.

just edit it ? but I just switch between `datafuse:master` and `datafuse:rust-2021`.

**Comment by drmingdrmer** [comment]
BTW, since `StateMachine::sm_tree` is a `SledTree` with key space support, you can just define a new `SledKeySpace` to impl a db or table. Just like `GenericKV` does:
```
pub struct GenericKV {}
impl SledKeySpace for GenericKV {
    const PREFIX: u8 = 6;
    const NAME: &'static str = "generic-kv";
    type K = String;
    type V = SeqValue;
}
```

**Comment by PsiACE** [comment]
edit-base

click `edit` and chose `base`.

**Comment by ariesdevil** [comment]
@PsiACE I think I was cheated by [this document](https://github.blog/changelog/2021-01-19-support-for-renaming-an-existing-branch/), it says:
> When a branch is renamed:

> Open pull requests and draft releases targeting the renamed branch will be retargeted automatically
Branch protection rules that explicitly reference the renamed branch will be updated

**Pull request closed without merging**